plugins/net_plugin/include/eosio/net_plugin/net_utils.hpp

```diff
@@ -40,29 +40,80 @@ namespace detail {
       return block_sync_rate_limit;
    }
 
-} // namespace detail
+   /// @return host, port, remainder
+   inline std::tuple<std::string, std::string, std::string> split_host_port_remainder(const std::string& peer_add) {
+      using std::string;
+      // host:port[:trx|:blk][:<rate>]
+      if (peer_add.empty()) return {};
 
-   /// @return listen address and block sync rate limit (in bytes/sec) of address string
-   inline std::tuple<std::string, size_t> parse_listen_address( const std::string& address ) {
-      auto listen_addr = address;
-      auto limit = std::string("0");
-      auto last_colon_location = address.rfind(':');
-      if( auto right_bracket_location = address.find(']'); right_bracket_location != address.npos ) {
-         if( std::count(address.begin()+right_bracket_location, address.end(), ':') > 1 ) {
-            listen_addr = std::string(address, 0, last_colon_location);
-            limit = std::string(address, last_colon_location+1);
-         }
-      } else {
-         if( auto colon_count = std::count(address.begin(), address.end(), ':'); colon_count > 1 ) {
+      string::size_type p = peer_add[0] == '[' ? peer_add.find(']') : 0;
+      if (p == 0) {
+         if( auto colon_count = std::count(peer_add.begin(), peer_add.end(), ':'); colon_count >= 7 ) {
             EOS_ASSERT( colon_count <= 2, chain::plugin_config_exception,
-                        "Invalid address specification ${addr}; IPv6 addresses must be enclosed in square brackets.", ("addr", address));
-            listen_addr = std::string(address, 0, last_colon_location);
-            limit = std::string(address, last_colon_location+1);
+                        "Invalid address specification ${a}; IPv6 addresses must be enclosed in square brackets.", ("a", peer_add));
          }
       }
+      string::size_type colon = p != string::npos ? peer_add.find(':', p) : string::npos;
+      if (colon == string::npos || colon == 0) {
+         return {};
+      }
+      string::size_type colon2 = peer_add.find(':', colon + 1);
+      string host = (p > 0) ? peer_add.substr( 0, p+1 ) : peer_add.substr( 0, colon );
```

Comment:
Wait, at this point, I think we are guaranteed that:
- `p > 0`
- `p + 1 == colon`

Reply:
I reworked the logic, please check again.